pkg/pool-utils/contracts/test/MockFailureModes.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/helpers/BalancerErrors.sol";
+
+abstract contract MockFailureModes {
+    enum FailureMode { INVARIANT, PRICE_RATE }
+
+    // Set to true to simulate a given failure mode
+    mapping(FailureMode => bool) private _failureState;
+
+    // Error message to return in simulated failure mode
+    mapping(FailureMode => uint256) private _failureCode;
+
+    modifier whenNotInFailureMode(FailureMode mode) {
+        _ensureNotFailed(mode);
+        _;
+    }
+
+    constructor() {
+        // Initialize error messages
+        _failureCode[FailureMode.INVARIANT] = Errors.STABLE_INVARIANT_DIDNT_CONVERGE;
+        _failureCode[FailureMode.PRICE_RATE] = Errors.LOW_LEVEL_CALL_FAILED;
```

Comment:
I'm torn on this. Intuitively I'd think it'd be better to error out with a custom message so that we can be sure that the cause for the failure was this module and not something inside the contracts. @TomAFrench wdyt?

Reply:
The idea was to mimic the behavior of production so that we could potentially reuse tests that already check for failures with the "real" messages, but I can see both sides.

It's actually flexible enough now for both. I could initialize all modes with a custom 'INDUCED_FAILURE' message - but if you wanted to override it (e.g., with STABLE_INVARIANT_DIDNT_CONVERGE, to reuse a test), you could. How about that?

Reply:
No, let's make it explicit that it is the mock contract that is failing. I don't really see many use cases where it'd make sense to reuse real tests.